src/za/ac/sun/cs/green/service/simplify/ConstantPropogation.java

```diff
@@ -0,0 +1,503 @@
+                expression.accept(simplifyVisitor);
+                expression = simplifyVisitor.getExpression();
+                changed = orderingVisitor.hasChanged() || 
+                    constantVisitor.hasChanged() ||
+                    simplifyVisitor.hasChanged();
+
+                if (n++ > 50) {
+                    log.log(Level.SEVERE, "Failed to simplify further after 50 attempts");
+                    break;
+                }
+
+            }
+
+			log.log(Level.FINEST, "After Propagation: " + expression);
+			return expression;
+		} catch (VisitorException x) {
+			log.log(Level.SEVERE,
+					"encountered an exception -- this should not be happening!",
+					x);
+		}
+		return null;
+	}
+
+	private static class OrderingVisitor extends Visitor {
+
+		private Stack<Expression> stack;
+        private boolean changed;
+
+		public OrderingVisitor() {
+			stack = new Stack<Expression>();
+            changed = false;
+		}
+
+		public boolean hasChanged() {
+			return changed;
+		}
+
+		public Expression getExpression() {
+			return stack.pop();
+		}
+
+		@Override
+		public void postVisit(IntConstant constant) {
+			stack.push(constant);
+		}
+
+		@Override
+		public void postVisit(IntVariable variable) {
+			stack.push(variable);
+		}
+
+		@Override
+		public void postVisit(Operation operation) throws VisitorException {
+			Operation.Operator op = operation.getOperator();
+
+			switch (op) {
+                case ADD:
+                case MUL:
+                    Expression r = stack.pop();
+                    Expression l = stack.pop();
+
+                    if (l instanceof IntConstant && r instanceof IntVariable) {
+                        stack.push(new Operation(op, r, l));
+                        changed = true;
+                    } else {
+                        stack.push(new Operation(op, l, r));
+                    }
+
+                    break;
+                default:
+                    int arity = operation.getOperator().getArity();
+                    Expression operands[] = new Expression[arity];
+                    for (int i = arity; i > 0; i--) {
+                        operands[i - 1] = stack.pop();
+                    }
+                    stack.push(new Operation(operation.getOperator(), operands));
+            }
+		}
+
+	}
+
+
+	private static class ConstantVisitor extends Visitor {
+
+		private Stack<Expression> stack;
+        private Map<IntVariable, IntConstant> constants;
+        private boolean replace;
+        private boolean changed;
+        private boolean unsatisfiable;
+
+		public ConstantVisitor(Map<IntVariable, IntConstant> constants) {
+			stack = new Stack<Expression>();
+            this.constants = constants;
+            replace = true;
+            changed = false;
+            unsatisfiable = false;
+		}
+
+		public boolean hasChanged() {
+			return changed;
+		}
+
+		public Expression getExpression() {
+            if (unsatisfiable) {
```

Comment:
Run a formatter over this file. The indentation makes following the uncommented mass even more difficult.